### PR TITLE
[MODULE] Add exercises and review DPO and ORPO module

### PR DESCRIPTION
**Pull request by burtenshaw**
This PR is to finalise module 2 for release on December 6.

### CODE REVIEW

File: 2_preference_alignment/orpo.md [hunk 1]
```diff
@@ -1,40 +1,25 @@
...
-The training process uses a preference dataset containing:
-- Input prompts
-- Chosen (preferred) responses
-- Rejected (disfavored) responses
-
-Unlike traditional methods that require separate stages and reference models, ORPO integrates preference alignment directly into the supervised fine-tuning process. This monolithic approach makes it:
-- Reference model-free
-- Computationally more efficient
-- Memory efficient (fewer FLOPs)
+Together, these components guide the model to adapt to desired generations for the specific domain while actively discouraging generations from the set of rejected responses. The odds ratio mechanism provides a natural way to measure and optimize the model's preference between chosen and rejected outputs. If you want to deep dive into the math, you can read the [ORPO paper](https://arxiv.org/abs/2402.01714). If you want to learn about ORPO from the implementation perspective, you should check out how loss for ORPO is calculated in the [TRL library](https://github.com/huggingface/trl/blob/main/trl/trainer/orpo_trainer.py#L660).
```
Comment (ariG23498) [inline 2_preference_alignment/orpo.md:22]: The link might change later. Better to use the permalink (https://github.com/huggingface/trl/blob/b02189aaa538f3a95f6abb0ab46c0a971bfde57e/trl/trainer/orpo_trainer.py#L660) What do you think?

**Review by ariG23498 — APPROVED** [review]
The readme looks good to me.
You can merge if you like, I am yet to try out the notebooks.

To unblock the PR, I would vote for merging this, and then add a separate PR later if something come up.

**Comment by burtenshaw** [comment]
Thanks.

File: 2_preference_alignment/dpo.md [hunk 2]
```diff
@@ -16,38 +16,57 @@ Next comes preference learning, where the model is trained on pairs of outputs -
 
 The core innovation of DPO lies in its direct optimization approach. Rather than training a separate reward model, DPO uses a binary cross-entropy loss to directly update the model weights based on preference data. This streamlined process makes training more stable and efficient while achieving comparable or better results than traditional RLHF.
 
+## DPO datasets
+
+Datasets for DPO are typically created by annotating pairs of responses as preferred or non-preferred. This can be done manually or using automated filtering techniques. Below is an example structure of a DPO dataset:
+
+| Input | Prompt | Chosen | Rejected |
```
Comment (alvarobartt) [inline 2_preference_alignment/dpo.md:23]: AFAIK input is not really required and may lead to confusion (?), I think following the standard prompt, chosen, rejected makes the most sense at least for single-turn, whilst for multi-turn you could use messages, chosen, and rejected instead too.

File: 2_preference_alignment/README.md [hunk 3]
```diff
@@ -12,35 +12,29 @@ Alternative approaches like ORPO combine instruction tuning and preference align
...
 ## Resources
+
 - [TRL Documentation](https://huggingface.co/docs/trl/index)
 - [DPO Paper](https://arxiv.org/abs/2305.18290)
 - [ORPO Paper](https://arxiv.org/abs/2402.01714)
 - [Argilla RLHF Guide](https://argilla.io/blog/mantisnlp-rlhf-part-8/)
-
+- [Blog post on DPO](https://huggingface.co/blog/dpo-trl)
+- [TRL example script on DPO](https://github.com/huggingface/trl/blob/main/examples/scripts/dpo.py)
+- [TRL example script on ORPO](https://github.com/huggingface/trl/blob/main/examples/scripts/orpo.py)
```
Comment (alvarobartt) [inline 2_preference_alignment/README.md:40]: Maybe link the handbook too? As it contains recipes based on TRL for popular fine-tunes such as Zephyr

```
- [Hugging Face Alignment Handbook](https://github.com/huggingface/alignment-handbook)
```

File: 2_preference_alignment/dpo.md [hunk 4]
```diff
@@ -16,38 +16,57 @@ Next comes preference learning, where the model is trained on pairs of outputs -
...
+You can find a collection of DPO datasets on Hugging Face [here](https://huggingface.co/collections/argilla/preference-datasets-for-dpo-656f0ce6a00ad2dc33069478).
+
 ## Implementation with TRL
 
-The Transformers Reinforcement Learning (TRL) library makes implementing DPO straightforward. Here's a basic example of setting up DPO training:
+The Transformers Reinforcement Learning (TRL) library makes implementing DPO straightforward. The `DPOConfig` and `DPOTrainer` classes follow the same `transformers` style API.
+Here's a basic example of setting up DPO training:
 
 ```python
 from trl import DPOTrainer
```
Comment (alvarobartt) [inline 2_preference_alignment/dpo.md:39]: Missing `DPOConfig` import here:

```python
from trl import DPOConfig, DPOTrainer
```

**Pull request merged**